handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,471 @@
+                throw new IllegalStateException();
+            }
+        };
+
+        handshake(latch, serverOcspContext, clientOcspContext, null, clientHandler);
+
+        // ATTENTION: This one is strange! It's a DecoderException caused by a NPE.
+        Throwable cause = causeRef[0];
+        assertTrue(cause instanceof DecoderException);
+    }
+
+    @Test(timeout = 10000L)
+    public void testServerException() throws Exception {
+
+        assumeTrue(OpenSsl.isOcspSupported());
+
+        final Throwable[] clientCauseRef = new Throwable[] { null };
+        final Throwable[] serverCauseRef = new Throwable[] { null };
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        ChannelInboundHandlerAdapter serverHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    serverCauseRef[0] = cause;
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    clientCauseRef[0] = cause;
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = new ServerOcspContext() {
+            @Override
+            public boolean optional() {
+                return false;
+            }
+
+            @Override
+            public byte[] staple() throws Exception {
+                throw new IllegalStateException();
+            }
+        };
+
+        ClientOcspContext clientOcspContext = new TestClientOcspContext(true);
+
+        handshake(latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+
+        // If the client rejects the staple then Client and Server will fail with
+        // Exceptions which contain these respective messages.
+        String clientMessage = clientCauseRef[0].getMessage();
+        assertTrue(clientMessage, clientMessage.contains("tlsv1 alert internal error"));
+
+        String serverMessage = serverCauseRef[0].getMessage();
+        assertTrue(serverMessage, serverMessage.contains("clienthello tlsext"));
+    }
+
+    private static void handshake(CountDownLatch latch, ServerOcspContext serverOcspContext,
+            ClientOcspContext clientOcspContext, ChannelHandler serverHandler,
+            ChannelHandler clientHandler) throws Exception {
+
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContext serverSslContext = SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(SslProvider.OPENSSL)
```

Comment:
can we run this test with `OPENSSL_REFCNT` too? We need to be careful about releasing in this case of course.